### DISCUSSION TITLE
curl turns 25 years old on March 20, 2023

### DISCUSSION THREAD

**Discussion by bagder** [curl/curl#10465] [General]
If you want to say congratulations, share a curl related memory or otherwise celebrate this historic event in text, this is the thread for you.

curl was first released on March 20, 1998 when it was uploaded and made available for the world. Back then it was 2,200 lines of code, it had 24 command line options and it supported only HTTP, GOPHER and FTP.

Exactly on curl's 25th birthday, we will ship curl 8.0.0.

[Online celebration](https://daniel.haxx.se/blog/2023/03/10/curl-25-years-online-celebration/) | [Fossified, curl 25 years](https://pod.fossified.com/2023/03/19/s01e02.html) |[Twenty-five years of curl](https://daniel.haxx.se/blog/2023/03/20/twenty-five-years-of-curl/) | [1998 - 2023 presentation](https://youtu.be/1aWpQ_t2qUw)

:champagne: :cake: :fireworks:

**Comment by cmeister2** [comment]
I'm happy to have been here for even a small part of curl's history, and to be responsible for some part of its success. Here's to many more years!

**Comment by Dammmien** [comment]
Congrats 🎉

It's impressive to see softwares still on top, that have stay up-to-date with new protocols or market standards and remain as popular as ever after so many years. Keep it up, it's inspiring.

And good job to all the contributors 👏🏼 !!

**Comment by trn1ty** [comment]
Congratulations. Here's hoping for another 25 years.

For a while I teased people with `curl http://some-link | sudo sh` written without context in the middle of my homepage. Meant to be a lesson to audit source code and verify against man-in-the-middle attacks lest someone drop too much salt in the soup (http without TLS and curling straight into `sh(1)` are both bad practice), it instead claimed probably 5-6 people's Linux installations with `rm -rf /` before I took it down, and made said individuals all somewhere between distraught and irate. As it turns out bold and underlined text is probably a better means of notification than indiscriminate file-directed violence.

**Comment by 3eka** [comment]
Congrats! Thanks four your efforts... Keep going!

**Comment by hornschorsch** [comment]
Thanks for a great tool.

I stumbled across curl i think more then 20 years ago, when looking for a tool to test https-connections. It was mentioned as a "nifty tool" i believe in a HOWTO that explained how to set up ssl certficates with apache.

I then used it in a website monitoring software that i developed and maintain until today. I even made a few patches that found their way into the official release and i wrote the first version of Curl::easy for perl. So curl has always been a companion to mee. Thanks again.

**Comment by bmercernccer** [comment]
Curl has gone from a small, specialized command line tool to a ubiquitous necessity that's now included in most operating systems around the world, while still maintaining quality and commitment to purpose. This is something to be very proud of. Congratulations and thanks!

**Comment by Albonycal** [comment]
Congratulations! I really appreciate all the work you all have put into making curl what it is <3

**Comment by jcamiel** [comment]
I would like to thank you and all the _many many_ contributors for giving us curl and especially __libcurl__!

Having a robust, maintained, reliable tool is awesome; allowing anybody to build anything with it is wonderful. 

Congrats, in advance, for the birthday, I'm looking forward for the next 25 years!

**Comment by tomcam** [comment]
Curl feels inevitable, like it always should have been there. One of the best pieces of software ever created. Thank you.

**Comment by LTS42** [comment]
Absolute legend.

**Comment by Hawzen** [comment]
curl FTW o7

**Comment by NicoJuicy** [comment]
I used curl for when a site builder went bankrupt, to scrape every of their clients and then contact them.

That was a very busy week.. 

Congratulations!

**Comment by bitops** [comment]
Thank you for all the hard work on a true workhorse tool. A curl command is guaranteed to be running somewhere at this very moment! It's surely saved many a project and been used in all kinds of creative ways the world will never know. 

🍻 to curl! "Du gamla, du fria!" 😄

**Comment by shearichard** [comment]
Thanks to everyone who has contributed to curl, a very useful tool, well done 👍.

**Comment by tortoisedoc** [comment]
When using curl life is just so much simpler. Back at F-secure, we had prolems with an SMTP component; a discussion arose how to best fix & test this; and curl got thrown in the mix. There was some disbelief in action as we realized that yes, curl actually supports SMTP! And very well!  Heres to 25 more years!

**Reply by mendhak** [reply]
My favourite part of curl is learning new things and features and capabilities about it. Like just now, here.

**Comment by JG24** [comment]
It must be an amazing feeling to work on software that is used by millions of programmers around the world. I wish you all the best!

**Comment by searls** [comment]
Congratulations, Daniel! I remember meeting you and being confused—you were so young, I'd always thought `curl` was one of those UNIX built-ins that was built at Berkeley or MIT back in the 60s/70s!

But don't let me give the impression 25 years isn't a really long time! Thank you for your service building such an invaluable tool!

**Comment by cvengler** [comment]
Thank you for this amazing tool!

cURL has always been a valuable source to me, ever since I started with network programming shortly before the pandemic. I began sending patches in around that time as well, which taught me a lot, not only in terms of technology, but also in terms of working in a project alongside other people.

I'm really thankful and in deep debt for everything that cURL has offered me so far (educational resources, thousands of successful HTTP transfers, the curl-up conferences, travel reimbursement for curl-up 2022, meeting awesome people @bagder, @icing, @captain-caveman2k, ...). For another 25 years! :partying_face:

**Comment by bitcynth** [comment]
I have used curl countless times, in part because I am certain that I have used the library many times without even knowing about it. It has been my default tool for testing APIs and otherwise poking at various HTTP(s) endpoints and even when playing with some more obscure protocols.

Thank you @bagder and all the other contributors for this incredibly useful tool and library!

**Comment by xquery** [comment]
A story then ... in Feb 1998 we created one of the first european gambling
websites ( in the form of littlewoods-pools.co.uk) which you can see in the
wayback archive


https://web.archive.org/web/19980211180828/http://www.littlewoods-pools.co.uk/

after the launch, prob April/May I had gone off to Wales for a break ...
during which the Saturday I got a call about 'things not working' and if I
could do anything about it ... it being update football results, 'spot the
ball' competition etc etc ... I had implemented a private cgi-bin (eg. a
large, very unique file name ... IPFS eat your heart out) just in case but
my custom client was only on my machine.

This is where my first encounter with curl eg. the mad scramble of 'no
computer' a university computer guy letting me use their network and me
being able to compile and use curl ... saved the day ... ever since libcurl
and curl have been in the toolbox ;)

Thanks to Daniel and all previous, current and future committers for their
passion, care, dare I say even love ... for this work is more public
service to mankind then anything else. Long may it live !

Jim Fuller

**Comment by crgimenes** [comment]
The curl is the _lingua franca_ when I want to demonstrate the workings of any of the APIs I've created. 
A genuinely essential tool.

**Comment by renanbastos93** [comment]
Awesome, I arrived later but I am very happy for CurL I use been every week, hahaha

**Comment by dscho** [comment]
Wow. What an accomplishment, what a demonstration of tenacity, leadership and commitment!

When I started contributing (on and off) to libcurl, it was in my role as Git for Windows maintainer. It was with a lot of trepidation that I submitted [my first big contribution](https://github.com/curl/curl/pull/1601) that would allow Git for Windows to allow selecting the SSL/TLS backend at runtime. I was afraid that I would make trivial mistakes, and sure enough, I did. The grace and patience with which the shortcomings of my contribution were met was very, very uplifting and inspiring.

It is also thanks to this experience that I always point to libcurl as one example of a well-run Open Source project whenever somebody asks me how to run a project.

In a larger context, it is amazing how ubiquitous cURL is. Not only in Git (of course! If you clone https://github.com/curl/curl, chances are that you use cURL itself to perform that action!) but really in so many projects. And in so many places. Even [on Mars](https://github.com/readme/featured/nasa-ingenuity-helicopter)!

![Mars 2020 Contributor Badge](https://github.githubassets.com/images/modules/profile/achievements/mars-2020-contributor-default.png)

With a large amount of gratitude, here's to the next 25 years: to infinity and beyond!

**Comment by augustohp** [comment]
🍻A toast for the next 25 years!

**Comment by fluffy-critter** [comment]
Congratulations on writing and maintaining this incredibly useful piece of software. I have used libCurl in so many of my projects, and it just keeps on getting better.

Curl is almost certainly what Randall Munroe had in mind when he drew [this evergreen comic](https://xkcd.com/2347/):

![](https://imgs.xkcd.com/comics/dependency.png)

Thank you for all the work that you do.

**Comment by askiiart** [comment]
Here's to more years, and even more planets!

**Comment by e1m1** [comment]
🥳🥳🥳🥳🎉🎉🎉🎉 absolutely essential software! Happy birthday, and cheers to all who maintain it

**Comment by sidcode** [comment]
Happy Birthday to a public good that has made lives so much livelier! Kudos to the craftsmanship of all the people who made it happen :)

**Comment by Candinya** [comment]
Thanks for the powerful & wonderful tool! 🥳 

My favorite http debugging method is running `curl -v http(s)://example.com` in command line and see the detailed information, like TLS handshakes, headers, response body etc. It's so powerful, elegant and light-weighted that in most situations I can immediately find out what happened to the origin server and then search for solutions. It has helped me a lot, and will help me more in the future I believe.

Happy birthday 🎂 🎉

**Comment by ryenus** [comment]
Interesting fact, the curl command is probably the only one that uses up all the letters from a to z for command line options, for some both lower and upper are used. Good luck to remember all of them! 😅

Happy birthday! 🎂🎂🎂 🎉🎉🎉

**Comment by benatkin** [comment]
huzzah!!!

**Comment by DHowett** [comment]
Happy birthday, curl! Here's to the community, and to @bagder! 🥂

**Comment by freekmurze** [comment]
Congratulations!! 🎉

**Comment by jeremiahlee** [comment]
After basic file system commands, curl was the first "real" command line program I used. It's been with me my entire professional career. Verbose mode taught me how HTTP and TLS worked in practice. Thank you @bagder and all the contributors!

**Comment by basisbit** [comment]
Happy birthday curl, and a huge thank you to all developers who helped make this project so successful!

**Comment by 1000283** [comment]
The proof that good software can last long and reach other planets. Keep it up!

**Comment by piccinato** [comment]
Happy curly birthday!

**Comment by bitcynth** [comment]
Happy 25th birthday curl 🎉

**Comment by soderlind** [comment]
Gratulerer med dagen! :birthday: :balloon: :tada:

**Comment by pheiduck** [comment]
Happy 25th Birthday curl!
It is always a pleasure to contribute to this open source project, CI maintenance and updates. 🎉🎂🎁🎊

**Comment by alesandroortiz** [comment]
Happy birthday curl! 🥳🎉 Congrats and many thanks to Daniel Stenberg and all the contributors over the years!

**Comment by recollir** [comment]
Happy Birthday. Grattis på födelsedagen. Hip hip, hurra, hurra, hurra, hurraaaaaa!

**Comment by 4mp3d** [comment]
Huge thanks and congrats. I love this tool

**Comment by nolessafool** [comment]
`curl` has materially and consistently changed and improved my life.  

I have always preferred the command line, and was a unix systems administrator professionally for 20 years;  curl quickly grew into being one the first things I installed on any system missing it (right after `ssh`, back near the beginning).  As someone who still administers systems as a volunteer, I continue to benefit like this.  I've tried wget, but I always come back to curl.

I thought I was already a power user, when I started my second career in tech support about 12 years ago.  Now, I use it *many times* daily at work, and often write public, customer-facing documentation about using it (for instance:  <https://docs.netlify.com/api/get-started/#make-a-request> and <https://answers.netlify.com/t/support-guide-understanding-and-debugging-prerendering/150#testing-the-service-with-your-content-2>).  

**I use it more than I use `ls` and `cd`, more than any other command except `git` - debugging the CDN and websites I help provide tech support on.**

Thank you for all the time, effort and love that have been put into one of the most critical pieces of software in my world, and here's to 25 more years!

**Comment by gmaxwell447** [comment]
I started using cURL (actuall, libcURL) back in 1996 with a startup company on an embedded product that used a closed ecosystem with a proprietary OS. After some porting adventures, cURL became an integral part of the product. We even contributed a bug fix! Since then, I'm always happy to see cURL pop up everywhere. This is such a great community, and although I don't directly develop with cURL, I'm still on the library mailing list and keep up with the progress of the project. Happy Anniversary!

**Comment by wenestvedt** [comment]
Thank you so much for building this amazing, foundational tool!

**Comment by azric** [comment]
Congratulations on having a product that is a quarter of a century young.  Not many products can make that claim!

**Comment by nickzman** [comment]
Congratulations on the 25th anniversary! Sorry I missed the party. Here's to another 25 years!

**Comment by stdevel** [comment]
Congratulations on the anniversary! Thanks a ton for your and the communities efforts to make `curl` unique - can't imagine not having it. Using it on a nearly daily basis.

Is there a recording from yesterdays event? Or slides from your presentation? I'm sure people would be interested in it.

All the best for the next 25 years!

**Reply by bagder** [reply]
See https://youtu.be/1aWpQ_t2qUw

**Reply by stdevel** [reply]
Thanks a lot for sharing!

**Comment by abhint** [comment]
Happy 25th birthday to curl!. I am grateful for the countless hours of work you have saved me over the years.
